tensor2tensor/data_generators/generator_utils.py

```diff
@@ -333,19 +333,18 @@ def generate():
     for source in sources:
       url = source[0]
       filename = os.path.basename(url)
-      read_type = "r:gz" if "tgz" in filename else "r"
-
       compressed_file = maybe_download(tmp_dir, filename, url)
 
-      with tarfile.open(compressed_file, read_type) as corpus_tar:
-        corpus_tar.extractall(tmp_dir)
-
       for lang_file in source[1]:
         tf.logging.info("Reading file: %s" % lang_file)
         filepath = os.path.join(tmp_dir, lang_file)
+        if not tf.gfile.Exists(filepath):
+          read_type = "r:gz" if filename.endswith("tgz") else "r"
+          with tarfile.open(compressed_file, read_type) as corpus_tar:
+            corpus_tar.extractall(tmp_dir)
```

Comment:
why did you move this down here into the loop? shouldn't it conceptually happen before the loop runs?

Reply:
This is intentional - it prevents extracting the tar, if it has been already extracted, so it saves time (if multiple Problems share the same tar).
More importantly, it is necessary if the input files are not available online, but I copied them to the `tmp` directory (and didn't want to pack them into tar, but I still want T2T to create the vocabulary).

Reply:
And now, reading my 6e825430c776 commit message, I remember a third reason: the files in the CzEng tar don't have write permission (-r--r--r--). So if you try to extract this tar file and it is already extracted, it will fail (and the whole t2t-datagen as well).